block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockMetadataStore.scala

```diff
@@ -0,0 +1,105 @@
+package coop.rchain.blockstorage.dag
+
+import cats.Monad
+import cats.effect.Sync
+import cats.mtl.MonadState
+import cats.syntax.all._
+import coop.rchain.models.BlockHash.BlockHash
+import coop.rchain.models.BlockMetadata
+import coop.rchain.shared.syntax._
+import coop.rchain.shared.{AtomicMonadState, Log}
+import coop.rchain.store.KeyValueTypedStore
+import monix.execution.atomic.AtomicAny
+
+import scala.collection.immutable.SortedMap
+
+object BlockMetadataStore {
+  def apply[F[_]: Sync: Log](
+      blockMetadataStore: KeyValueTypedStore[F, BlockHash, BlockMetadata]
+  ): F[BlockMetadataStore[F]] =
+//    implicit val raiseIOError: RaiseIOError[F] = IOError.raiseIOErrorThroughSync[F]
+    for {
+      // Migrating data from previous metadata index
+//      result <- BlockMetadataPersistentIndex.read(Paths.get("index_path"))
+//      (oldMetadaatMap, _, _) = result
+//      _ <- blockMetadataStore.put(oldMetadaatMap.toSeq)
+      blockMetadataMap <- blockMetadataStore.toMap
+      dagState         = recreateInMemoryState(blockMetadataMap)
+    } yield new BlockMetadataStore[F](
+      blockMetadataStore,
+      new AtomicMonadState(AtomicAny(dagState))
+    )
+
+  private final case class DagState(
+      dagSet: Set[BlockHash],
+      childMap: Map[BlockHash, Set[BlockHash]],
```

Comment:
I see this notion of a `child` in many places. But aren't we actually more interested in parents not children?

Reply:
I see it's used in Estimator and LastFinalizedBlockCalculator. Do you think this can be rewritten in different way. Or maybe it's easy calculate childs from parents where is needed.

Reply:
> I see it's used in Estimator and LastFinalizedBlockCalculator. Do you think this can be rewritten in different way. Or maybe it's easy calculate childs from parents where is needed.

I'm not sure if there is a good way, lets better keep it as it is now.. We have parents in BlockMetadata, dagState is just  additional data, so my question is not relevant, this has been added fo particular usecase.